libraries/chain/webassembly/crypto.cpp

```diff
@@ -100,4 +106,143 @@ namespace eosio { namespace chain { namespace webassembly {
+      if(res.first != fc::alt_bn128_error::none) {
+         return error_code::fail;
+      }
+
+      *result = res.second;
+      return error_code::none;
+   }
+
+   int32_t interface::mod_exp(span<const char> base,
+                              span<const char> exp,
+                              span<const char> modulus,
+                              span<char> out) const {
+      using error_code = eosio::chain::webassembly::error_codes::crypto;
+
+      bytes bbase(base.data(), base.data() + base.size());
+      bytes bexp(exp.data(), exp.data() + exp.size());
+      bytes bmod(modulus.data(), modulus.data() + modulus.size());
+
+      auto res = fc::modexp(bbase, bexp, bmod);
+      if(res.first != fc::modular_arithmetic_error::none) {
+         return error_code::fail;
+      }
+
+      if( out.size() < res.second.size() )
+         return error_code::fail;
+
+      std::memcpy( out.data(), res.second.data(), res.second.size() );
+      return error_code::none;
+   }
+
+   int32_t interface::blake2_f( uint32_t rounds,
+                                span<const char> state,
+                                span<const char> message,
+                                span<const char> t0_offset,
+                                span<const char> t1_offset,
+                                bool final,
+                                span<char> out) const {
+
+      using error_code = eosio::chain::webassembly::error_codes::crypto;
+
+      bytes bstate(state.data(), state.data() + state.size());
+      bytes bmessage(message.data(), message.data() + message.size());
+      bytes bt0_offset(t0_offset.data(), t0_offset.data() + t0_offset.size());
+      bytes bt1_offset(t1_offset.data(), t1_offset.data() + t1_offset.size());
+
+      auto checktime = [this]() { context.trx_context.checktime(); };
+
+      auto res = fc::blake2b(rounds, bstate, bmessage, bt0_offset, bt1_offset, final, checktime);
+      if(res.first != fc::blake2b_error::none) {
+         return error_code::fail;
+      }
+
+      if( out.size() < res.second.size() )
+         return error_code::fail;
+
+      std::memcpy( out.data(), res.second.data(), res.second.size() );
+      return error_code::none;
+   }
+
+   void interface::sha3( span<const char> input, span<char> output, bool keccak ) const {
+      const size_t bs = eosio::chain::config::hashing_checktime_block_size;
+      const char* data = input.data();
+      uint32_t datalen = input.size();
+      fc::sha3::encoder enc{keccak};
+      while ( datalen > bs ) {
+         enc.write( data, bs);
+         data    += bs;
+         datalen -= bs;
+         context.trx_context.checktime();
+      }
+      enc.write( data, datalen);
+      auto res = enc.result();
+
+      auto copy_size = std::min( output.size(), res.data_size() );
+      std::memcpy( output.data(), res.data(), copy_size );
+   }
+
+   int32_t interface::ecrecover( span<const char> signature, span<const char> digest, span<char> pub) const {
```

Comment:
Feels like a missed opportunity to add this without support for EOS key types. Such functionality has long been requested.

Reply:
If any of these could be useful for non-EVM specific use cases, like you're suggesting here and I suspect others would also be useful as well, then might it make sense to give this feature a more generic/feature specific name and not something that suggests it's only useful for EVM implementations?

Reply:
That's a good point. These arguably should have a more generic protocol feature name because some of the new host functions are certainly not specific to EVM: anyone might want to make a sha3 hash for all sorts of purposes.